sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java

```diff
@@ -205,4 +231,83 @@ private Query query(FieldDoc lastDoc) {
         searchAfterQuery.add(optimizedQuery, BooleanClause.Occur.MUST_NOT);
         return searchAfterQuery.build();
     }
+
+    private static void raiseIfKilled(@Nullable Throwable t) {
+        if (t != null) {
+            Exceptions.rethrowUnchecked(t);
+        }
+    }
+
+    private static class KillableCollector implements Collector, Killable {
+
+        private final Collector delegate;
+
+        @Nullable
+        private volatile KillableLeafCollector currentLeafCollector;
+
+        @Nullable
+        private volatile Throwable killed;
+
+        public KillableCollector(Collector delegate, @Nullable Throwable killed) {
+            this.delegate = delegate;
+            this.killed = killed;
+        }
+
+        @Override
+        public LeafCollector getLeafCollector(LeafReaderContext context) throws IOException {
+            raiseIfKilled(killed);
+            currentLeafCollector = new KillableLeafCollector(delegate.getLeafCollector(context), killed);
+            return currentLeafCollector;
+        }
+
+        @Override
+        public ScoreMode scoreMode() {
+            return delegate.scoreMode();
+        }
+
+        @Override
+        public void kill(@Nonnull Throwable t) {
+            if (killed != null) {
+                return;
+            }
+            killed = t;
+            var leafCollector = currentLeafCollector;
+            if (leafCollector != null) {
```

Comment:
I think there is a race condition between getLeafCollector and kill:

```
t1: Is before `killed = t;`
t2: Is past `raiseIfKilled(killed);` but hasn`t done the assignment to `currentLeafCollector`
```

In that case the kill would be swallowed I think. Could probably be solved with a compare and set operation

Reply:
I don't think so as `killed` is passed to the leaf collector ctor.

Reply:
Yes, but only if `killed` has been assigned before the leaf collector is created.

The scenario I have in mind is that `killed` hasn't been assignd to yet, and the other thread read the volatile (unassigned) 'killed' but didn't finish the 'currentLeafCollector'  assignment yet.

How about having a final AtomicReference in the top most class and the collector + leaf collector receive a reference to that? That would replace the individual killed attributes.

Reply:
Right, there is still a race condition, good catch.

> How about having a final AtomicReference in the top most class and the collector + leaf collector receive a reference to that? That would replace the individual killed attributes.

Yep, sounds good.

Reply:
Pushed a fixup.